trace-forward/src/Trace/Forward/Network/Acceptor.hs

```diff
@@ -0,0 +1,200 @@
+            (AcceptedConnectionsLimit maxBound maxBound 0)
+            address
+            unversionedHandshakeCodec
+            timeLimits
+            (cborTermVersionDataCodec unversionedProtocolDataCodec)
+            acceptableVersion
+            (simpleSingletonVersions
+              UnversionedProtocol
+              UnversionedProtocolData
+              (SomeResponderApplication app))
+            nullErrorPolicies
+            $ \_ serverAsync -> wait serverAsync -- Block until async exception.
+
+acceptTraceObjects
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  => AcceptorConfiguration lo
+  -> ([lo] -> IO ())
+  -> (NodeInfo -> IO ())
+  -> RunMiniProtocol 'ResponderMode LBS.ByteString IO Void ()
+acceptTraceObjects config loHandler niHandler =
+  ResponderProtocolOnly $
+    MuxPeerRaw $ \channel ->
+      runPeer
+        (acceptorTracer config)
+        (Acceptor.codecTraceForward CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode)
+        channel
+        (Acceptor.traceAcceptorPeer $
+          acceptorActions config loHandler niHandler True False)
+
+acceptTraceObjectsInit
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  => AcceptorConfiguration lo
+  -> ([lo] -> IO ())
+  -> (NodeInfo -> IO ())
+  -> RunMiniProtocol 'InitiatorMode LBS.ByteString IO () Void
+acceptTraceObjectsInit config loHandler niHandler =
+  InitiatorProtocolOnly $
+    MuxPeerRaw $ \channel ->
+      runPeer
+        (acceptorTracer config)
+        (Acceptor.codecTraceForward CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode)
+        channel
+        (Acceptor.traceAcceptorPeer $
+          acceptorActions config loHandler niHandler True False)
+
+acceptorActions
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  => AcceptorConfiguration lo -- ^ Acceptor's configuration.
+  -> ([lo] -> IO ())          -- ^ The handler for accepted 'TraceObject's.
+  -> (NodeInfo -> IO ())      -- ^ The handler for accepted info about the node.
+  -> Bool                     -- ^ The flag for node's info request: only once in the beginning.
+  -> Bool                     -- ^ The flag for end of session: if 'True', 'SendMsgDone' will be sent.
+  -> Acceptor.TraceAcceptor lo IO ()
+acceptorActions config@AcceptorConfiguration{whatToRequest, shouldWeStop} loHandler niHandler askForNI False =
+  -- We are able to send request for:
+  -- 1. node's info,
+  -- 2. new 'TraceObject's.
+  -- But request for node's info should be sent only once (in the beginning of session).
+  if askForNI
+    then
+      Acceptor.SendMsgNodeInfoRequest $ \replyWithNI -> do
+        runActionWithTimer (niHandler replyWithNI) limitInSecs
+        readIORef shouldWeStop <&> acceptorActions config loHandler niHandler False
```

Comment:
This is not do what I asked for.  I wanted the whole `acceptActions` to terminate if the `IORef` is set to true and does not terminate within a timeout, now it will timeout the `niHandler` action but it has nothing to do with the `IORef`.

Reply:
I'd suggest to turn the `IORef` into a `TVar` and use it in `runActionWithTimer`.  It will require two stm transactions first would block on on this tvar. Then one can create a timerVar and block on it, something like this:
```
timeoutWhenStopped :: TVar Bool -> Int ->  IO a -> IO (Maybe a)
timeoutWhenStoppepd stopVar delay io =
  fmap (either (\_ -> Nothing) Just) $
  race io
         $ do atomically (readTVar stopVar >>= check)
                 v <- registerDealay delay    
             atomically (readTVar v >>= check)       
```
You cannot apply it to `TraceAcceptor` type directly, since it's not an `IO` action, but you can plug it where you run the acceptor using `runPeer`.

Reply:
@coot, what you are alluding to is quite involved -- can you suggest a concrete change -- we'll be happy to accept.

Reply:
The biggest change is `IORef -> TVar`, but it's a mechanical change - so quite easy to do.  Other than that it's just wrapping `runPeer` inside `timeoutWhenStopped`.

Reply:
`TVar` instead of `IORef` is already done.

Reply:
`TVar` instead of `IORef` is already done.

Reply:
`TVar` instead of `IORef` is already done: https://github.com/input-output-hk/cardano-node/blob/c6f8f7555993cf31c8a63906ef42ce4dad6a6118/trace-forward/src/Trace/Forward/Network/Acceptor.hs#L168

Reply:
> wrapping `runPeer` inside `timeoutWhenStopped`

Honestly, I cannot understand why?? I do understand why we need stop timer - to avoid too long (theoretically - infinite) time for performing of external handlers. And now it's STM-transactions with `orElse` - ok, I understand. But why do we need to wrap `runPeer` inside `timeoutWhenStopped`? `runPeer` calls `acceptorActions`, `acceptorActions` calls `runActionWithTimer`...

Reply:
`Peer`, which is a result of calling `acceptorActions` is just a data type, you cannot effectively limit how long it will execute after the stop var was set to `True`, or any other resource handling.  You can only do that around `runPeer` which returns an `IO` action, `IO` actions can be timed out.

Reply:
Ok. But I still don't understand how to do it exactly. Are there any existing typed-protocols in `ouroboros-network` where I can find an example of it? And does any from existing protocols have such a `TVar`-based stopping mechanism?